### PR TITLE
[DebugInfo] Add DW_OP_LLVM_extract_bits

### PR DESCRIPTION
**Pull request by john-brawn-arm**
This operation extracts a number of bits at a given offset and sign or zero extends them, which is done by emitting it as a left shift followed by a right shift.

This is being added for use in clang for C++ structured bindings of bitfields that have offset or size that aren't a byte multiple. A new operation is being added, instead of shifts being used directly, as it makes correctly handling it in optimisations (which will be done in a later patch) much easier.

### CODE REVIEW

**Review by adrian-prantl — COMMENTED** [review]
I like this approach for its readability and because it's a lot simpler than mixing this up with the composition operators.

Can you explain why the signedness matters?

File: llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+!12 = !{i32 2, !"Debug Info Version", i32 3}
+!13 = distinct !DISubprogram(name: "test1", linkageName: "test1", scope: !3, file: !3, line: 8, type: !14, scopeLine: 8, flags: DIFlagPrototyped, spFlags: DISPFlagDefinition, unit: !2, retainedNodes: !16)
+!14 = !DISubroutineType(types: !15)
+!15 = !{!9}
+!16 = !{}
+!17 = !DILocalVariable(name: "x", scope: !13, file: !3, line: 9, type: !9)
+!18 = !DILocation(line: 9, column: 9, scope: !13)
+!19 = !DILocalVariable(name: "y", scope: !13, file: !3, line: 9, type: !20)
+!20 = !DIBasicType(name: "signed int", size: 32, encoding: DW_ATE_signed)
+!21 = !DILocation(line: 9, column: 12, scope: !13)
```
Comment (adrian-prantl) [inline llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll:86]: You an probably collapse most of the DILocations to a single line 0 one to make the example more focussed.

File: llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  %0 = load i8, ptr @g, align 1
+  tail call void @llvm.dbg.value(metadata i8 %0, metadata !24, metadata !DIExpression(DW_OP_LLVM_extract_bits, 0, 3, DW_ATE_unsigned)), !dbg !25
+  tail call void @llvm.dbg.value(metadata i8 %0, metadata !26, metadata !DIExpression(DW_OP_LLVM_extract_bits, 3, 4, DW_ATE_signed)), !dbg !27
+  ret i8 %0, !dbg !28
+}
+
+; CHECK-LABEL: DW_TAG_subprogram
+; CHECK: DW_AT_name ("test3")
+; CHECK: DW_TAG_variable
+; CHECK: DW_AT_location (DW_OP_breg0 RAX+0, DW_OP_constu 0x3f, DW_OP_shr, DW_OP_stack_value)
```
Comment (adrian-prantl) [inline llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll:45]: This test seems to hardcode some register allocation assumptions. What do you think about converting it to MIR?

Comment (john-brawn-arm) [reply llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll:45]: It looks like MIR doesn't handle multiple debug values in the same stack location. The MIR generated for test1 (by running DW_OP_LLVM_extract_bits.ll through llc with --stop-after=patchable-function) has
```
stack:
  - { id: 0, name: '', type: default, offset: -9, size: 1, alignment: 1,
      stack-id: default, callee-saved-register: '', callee-saved-restored: true,
      debug-info-variable: '!17!19', debug-info-expression: '!DIExpression(DW_OP_LLVM_extract_bits, 0, 3, DW_ATE_unsigned)!DIExpression(DW_OP_LLVM_extract_bits, 3, 4, DW_ATE_signed)',
      debug-info-location: '!18!21' }
```
which llc then gives an error for on reading it:
```
error: tmp.mir:130:32: expected end of string after the metadata node
      debug-info-variable: '!17!19', debug-info-expression: '!DIExpression(DW_OP_LLVM_extract_bits, 0, 3, DW_ATE_unsigned)!DIExpression(DW_OP_LLVM_extract_bits, 3, 4, DW_ATE_signed)',
                               ^
```

**Comment by john-brawn-arm** [comment]
> Can you explain why the signedness matters?

If we have
```
typedef struct __attribute__((packed)) {
  unsigned int x : 4;
  signed int y : 4;
} struct_t;

struct_t g = {1, -1};

int main() {
  auto [x, y] = g;
  return x+y;
}
```
then if y isn't sign-extended, looking at the values in a debugger we have
```
    frame #0: 0x00007ffff7fff164 a.out`main at tmp.cpp:10:10
   7
   8    int main() {
   9      auto [x, y] = g;
-> 10     return x+y;
   11   }
(lldb) p x
(unsigned int) 1
(lldb) p y
(int) 15
(lldb) p x+y
(unsigned int) 16
```
which is wrong. If it's sign-extended we have
```
    frame #0: 0x00007ffff7fff164 a.out`main at tmp.cpp:10:10
   7
   8    int main() {
   9      auto [x, y] = g;
-> 10     return x+y;
   11   }
(lldb) p x
(unsigned int) 1
(lldb) p y
(int) -1
(lldb) p x+y
(unsigned int) 0
```
which is correct.

File: llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll (hunk 1 above)
Comment (john-brawn-arm) [reply llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll]: Done.

**Comment by adrian-prantl** [comment]
I see: the value of the 4-bit signed field is `-1`, which is represented as `1111` in binary. And if there are other arithmetic DWARF operations after that would cause a problem for the untyped DWARF stack, where that value would be interpreted as 15 decimal. I think it's reasonable to have signedness bit in the operation under these circumstances. It would be more space-efficient to introduce two separate operations in in `u` and `s` variants, but I don't have any strong opinion either way.

I think I'm fine with this proposal! It would be good to hear at least one more LGTM from others in the @llvm/pr-subscribers-debuginfo group before landing this.

**Comment by OCHyams** [comment]
> as it makes correctly handling it in optimisations (which will be done in a later patch) much easier.

In case it's relevant to you have you seen #61981? Variable locations for structured bindings are currently broken if the alloca gets SROA'd. I've got a private patch that needs some reworking that I've not had time to put upstream. I think there are other issues too (I can't recall the details, but I think it's related to converting address offsets to bit-shifts after mem2reg, or the lack thereof).

---

I think it makes sense to add "high level" DIExpression operations like this and this one seem reasonable. I'd feel more confident about the patch if I could see the whole stack, but github does make that somewhat difficult (perhaps you could share a link to a branch with all the patches you plan on submitting, if it's not too much of a faff?).

LGTM from me too but I'd prefer to wait a few days before accepting to see if anyone else wants to chip in.

File: llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll (hunk 2 above)
Comment (SLTozer) [reply llvm/test/DebugInfo/X86/DW_OP_LLVM_extract_bits.ll:45]: Hm, this is a silly bug - I remember seeing it some time ago, I thought it had been fixed! This isn't a technical limitation of MIR in-memory, but an error in how it is printed (and possibly parsed) that should be fixed soon; I'll take a look at it soon. In the meanwhile, you could use regex (something like `{{R[^+]+}}`?) to match any 64 bit register.

**Comment by SLTozer** [comment]
> It would be more space-efficient to introduce two separate operations in in `u` and `s` variants, but I don't have any strong opinion either way.

I agree with this, we aren't squeezed for `DW_OP_LLVM_` operand space right now, but we care do care about the size of DIExpressions, so splitting into signed/unsigned variants would be better imo, unless there's a specific advantage to making the encoding an operand instead.

**Review by SLTozer — COMMENTED** [review]
Broadly LGTM, with the question about the operator encoding outstanding.

**Review by SLTozer — APPROVED** [review]
Changes LGTM.

**Comment by john-brawn-arm** [comment]
I've put draft versions of the changes I'll be making in clang in #94632 and for SROA support in #94638.

**Pull request merged**